### ISSUE TITLE
Thoughts on array parallelism

### ISSUE THREAD

**Issue by andrewdavidmackenzie** [andrewdavidmackenzie/flow#199]
Inputs/Outputs of a process can be:
- nothing
- single value
- series of input values (depth)   (not on output)
- array of values

Combinations:

 0)                           -> P -> 
 1)                           -> P -> Single value
 2)                           -> P -> Array of values

 3) Single value       -> P
 4) Single value       -> P -> Single value
 5) Single value       -> P -> Array of values

 6) Series of values -> P
 7) Series of values -> P -> Single value
 8) Series of values -> P -> Array of values

 9) Array of values  -> P
10) Array of values  -> P -> Single value
11) Array of values  -> P -> Array of values

If the input to a process is not an array, but an array is being sent to it, then we could spawn multiple instances to handle each element of the array.

The need to combine the output values will depend on that the process output is, a single value (each instance do it) or if it's an array (we need to re-combine outputs from the multiple instances).

**Comment by andrewdavidmackenzie** [comment]
When a function can output a number of values, have a way to allow it to output them in series, or in parallel?

e.g. a range of numbers....
Series: one is output at a time, on each invocation.

Parallel: output a range in one go and the system (maybe) divides them between multiple receivers. If receivers receive a range of numbers one by one will depend on their definition, but a range would be a different data type (array of numbers).

Maybe, if a function outputs an array of something, be able to specify next function as a serial processor, or a "fan out" that would split it into many individual numbers that maybe spawns many runnables in parallel (depends on the runtime how many, from 1 up).

**Comment by andrewdavidmackenzie** [comment]
When an array output from one process is connected to an input of a single element on another process. When the array of values is produced:
- we produce a stream of values from the array

One process could be run per value generated, or a number of parallel processes run to take a value of the stream each in parallel.

This case shouldn't change the parallelisation rules at all really, except that the production of the Array of output allows you to see more easily that multiple values are available, and multiple processes can be run in parallel on them.

**Comment by andrewdavidmackenzie** [comment]
If process A produces one output value, and downstream process B accepts that one value:
- when a number of Process A are run in parallel, then a number of process B can also be run in parallel, each processing the output of a single Process A

If downstream process B accepts an array of values, and the runtime is responsible for accumulating them, then it must do so from multiple instances of Process A, and accumulate them for a process B instance. If only one was accumulated for many instances of Process B then it could never run.
-> Accumulate on the input of process B until enough to run it, and do so, and continue to accumulate on the input. If enough are ready in time, them run a second parallel instance of process B with the second batch.

**Comment by andrewdavidmackenzie** [comment]
Take processes that produce series (like series, or fibonacci):

When we have parallel execution of them, how to enforce the order?
Will need:
- to know when order has ended.Some for of EOF symbol in a stream?
- have an ordering process that gathers (reduces) the outputs and orders them?

Process can declare if it can run in parallel or not?
Would prefer to avoid that and make it implicit.

Best would be that the process is defined in such a way as to produce a series, and that parallel execution cannot work - due to data dependencies....

**Comment by andrewdavidmackenzie** [comment]
Currently an input can define a "depth" such that it needs 'n' elements to be considered "filled" and the process to run.

Consider defining this so that the process's input is an array of a given width (of the same element type) and the runtime is responsible for accumulating values from the stream of single values until an array can be created and provided as input.

Array input can be of:

fixed width - must be met for input to fill
variable - if an array is sent, all elements will be processed together - but this cannot convert from stream of elements as we don't know how many to accumulate.
Is more a cosmetic change, and how the process defines it's input - that a big implementation change.

Document array input from a stream of single values.

**Comment by andrewdavidmackenzie** [comment]
If a process produces an array of elements, then allow it to be connected to a process that accepts a single element of that type - and have the runtime take the output array and send to the received one by one.

**Comment by andrewdavidmackenzie** [comment]
Implemented
